source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.h

```diff
@@ -0,0 +1,150 @@
+
+/**
+ * A concurrency controller that implements the a variation of the Gradient algorithm to control the
+ * allowed concurrency window.
+ *
+ * An ideal round-trip time (minRTT) is measured periodically by only allowing a single outstanding
+ * request at a time and measuring the round-trip time to the listener. This information is then
+ * used in the calculation of a number called the gradient, using time-sampled latencies
+ * (sampleRTT):
+ *
+ *     gradient = minRTT / sampleRTT
+ *
+ * This gradient value has a useful property, such that it decreases as the sampled latencies
+ * increase. The value is then used to periodically update the concurrency limit via:
+ *
+ *     limit = old_limit * gradient
+ *     new_limit = limit + headroom
+ *
+ * The headroom value allows for request bursts and is also the driving factor behind increasing the
+ * concurrency limit when the sampleRTT is in the same ballpark as the minRTT. This value must be
+ * present in the calculation, since it forces the concurrency limit to increase until there is a
+ * deviation from the minRTT latency. In its absence, the concurrency limit could remain stagnant at
+ * an unnecessarily small value if sampleRTT ~= minRTT. Therefore, the headroom value is
+ * unconfigurable and is set to the square-root of the new limit.
+ */
+class GradientController : public ConcurrencyController {
+public:
+  GradientController(GradientControllerConfigSharedPtr config, Event::Dispatcher& dispatcher,
+                     Runtime::Loader& runtime, const std::string& stats_prefix,
+                     Stats::Scope& scope);
+
+  ~GradientController() override;
+
+  // ConcurrencyController.
+  RequestForwardingAction forwardingDecision() override;
+  void recordLatencySample(const std::chrono::nanoseconds& rq_latency) override;
+  int concurrencyLimit() const override { return concurrency_limit_.load(); }
+
+private:
+  static GradientControllerStats generateStats(Stats::Scope& scope,
+                                               const std::string& stats_prefix);
+  void updateMinRTT();
+  std::chrono::microseconds processLatencySamplesAndClear()
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(latency_sample_mtx_);
+  int calculateNewLimit() ABSL_EXCLUSIVE_LOCKS_REQUIRED(update_window_mtx_);
+  void setMinRTTSamplingWindow() ABSL_EXCLUSIVE_LOCKS_REQUIRED(update_window_mtx_);
+  void resetSampleWindow() ABSL_EXCLUSIVE_LOCKS_REQUIRED(update_window_mtx_);
+
+  GradientControllerConfigSharedPtr config_;
+  Event::Dispatcher& dispatcher_;
+  Stats::Scope& scope_;
+  GradientControllerStats stats_;
+
+  absl::Mutex update_window_mtx_;
+  std::chrono::nanoseconds min_rtt_;
+  std::atomic<bool> recalculating_min_rtt_;
+  absl::Mutex latency_sample_mtx_ ABSL_ACQUIRED_AFTER(update_window_mtx_);
+  std::chrono::nanoseconds sample_rtt_ ABSL_GUARDED_BY(update_window_mtx_);
+
+  std::atomic<int> num_rq_outstanding_;
+  std::atomic<int> concurrency_limit_;
```

Comment:
From a high level view of the code, it's a little hard to understand the threading model here. Two things:
1) Please only use atomic variables if they are actually needed and the values are not otherwise checked within a lock. This might be already the case but it's not completely clear so maybe more comments.
2) Do we definitely need two locks? It would be better to avoid that unless there is a good reason. If so, more comments please.